### DISCUSSION TITLE
[Craft 4] Change order of defaultTemplateExtensions

### DISCUSSION THREAD

**Discussion by ostark** [craftcms/cms#11809] [Ideas] [answered]
Craft 4 uses `.twig` as a file extension for internal templates and recommends it in the docs for site templates as well.

It would make sense to change the order to look them up first: https://github.com/craftcms/cms/blob/main/src/config/GeneralConfig.php#L814

```php
public array $defaultTemplateExtensions = ['twig', 'html'];
```

**Comment by brandonkelly** [comment, marked as answer]
Good point, however that could be a breaking change since if anyone has `.twig` and `.html` templates of the same name, so I’ve made the swap for Craft 5 rather than 4.x. (3ca636eaae15cc5ec6e3aa15b4be08a6071abc7d)

**Reply by ostark** [reply]
That was actually the reason!
When upgrading to Craft 4 using non-atomic deployments (rsync / sftp) that don't delete old files creates an odd state, since Craft will pick the old .html version instead of the new .twig (same file name).

**Reply by brandonkelly** [reply]
That feels like a deployment problem. Craft should not have to worry about other files polluting its source directories, even if those other files are from prior versions of itself.

**Reply by ostark** [reply]
@brandonkelly Sure this deployment way has disadvantages. I'm totally fine with your answer! (I thought)

Then I looked at the code on how the setting is used and noticed that users are never in control of the loading order of the CP templates, it's hardcoded. With the change from .html to .twig in Craft 4, you create some I/O stat operations on a non-existing `.html` file for CP rendering, if I read the code correctly.

https://github.com/craftcms/cms/blob/71aac29fa480dd3b09a72c4fa54a5fd4976b4dd4/src/web/View.php#L1465

It also means, there is no way for us (or the user) to adjust it via ENV var or plugin or config, which is sad.
The setting affects frontend templates only, which means your change on the 5.x branch is kind of useless. I can not see a serious breaking change for users if you would change it in `View`.

**Reply by brandonkelly** [reply]
> With the change from .html to .twig in Craft 4, you create some I/O stat operations on a non-existing .html file for CP rendering, if I read the code correctly.

True. I’ve just fixed that for Craft 4.3 by explicitly referencing template paths including their extension, and I’ve also changed the order of extension preference for control panel templates only, which still runs the risk of being a breaking change for plugins, but the changes are much more slim. (#11840)

> It also means, there is no way for us (or the user) to adjust it via ENV var or plugin or config, which is sad.

It wouldn’t make any sense for that to be configurable for control panel templates. The whole reason that the supported extensions (and index template filenames) are hard-coded for the control panel is so Craft and plugin templates can safely stick with the `index`/`.html`/`.twig` conventions and not have to worry about them being configured differently on some environments.

**Reply by brandonkelly** [reply]
Worth mentioning that there’s still plenty of room for similar issues to crop up for front-end templates though, where developers may want to follow suit and start renaming their own templates to `.twig` without thinking about changing the config setting. So worth looking into improving the deployment pipeline regardless of the Craft 4.3/5.0 changes, if possible.

**Reply by ostark** [reply]
Thanks, I do appreciate the change.
